Generalize to no longer be SARS-CoV-2 specific

terrycojones opened the pull request:
Generalized to deal with non-SARS2 features. Added annotate-genome.py script. Handle translation of reverse-complemented ORFs. Allow passing names of translated features and also feature aliases. Make it possible to load features from a JSON file (as produced by annotate-genome.py).  Added checking for multiple genome ranges. Made it possible to treat unannotated genome regions as regular features.

Also, the code has been reformatted to use [black](https://black.readthedocs.io/en/stable/index.html)

terrycojones committed:
Corrected some tiny black issues.

terrycojones committed:
Generalized to deal with non-SARS2 features. Added annotate-genome.py script. Handle translation of reverse-complemented ORFs. Allow passing names of translated features and also feature aliases. Make it possible to load features from a JSON file (as produced by annotate-genome.py).  Added checking for multiple genome ranges. Made it possible to treat unannotated genome regions as regular features.

bamueh commented on bin/annotate-genome.py:
```diff
@@ -0,0 +1,195 @@
+                    file=sys.stderr,
+                )
+
+    return 0
```
Isn't that the default?

terrycojones replied:
No. Functions return `None` by default.

bamueh commented on bin/annotate-genome.py:
```diff
@@ -0,0 +1,195 @@
+        action="store_true",
+        help=(
+            "If the genome ORF differs from the reference due to no start "
+            "or stop codon, print a message.",
```
Do you really need this trailing comma?

terrycojones replied:
Huh - thanks! I guess that help string was shorter once and had a comma after it. then I made it longer and put the parens in wrongly and accidentally created a `tuple`.  I've started trying to use typing hints and checking the code with `mypy` but there are still too many warnings to deal with so I'm going slowly and not running `mypy` all the time. Had I done so, I would perhaps have seen that it says:

```
bin/annotate-genome.py:184: error: Argument "help" to "add_argument" of "_ActionsContainer" has incompatible type "Tuple[str]"; expected "Optional[str]"
```

Anyway, thanks :-)

terrycojones committed:
Removed trailing comment that was making a help string out of a tuple.

terrycojones committed:
Make annotate-genome.py only print differences in its summary.

terrycojones commented:
I changed `annotate-genome.py` to (if `--summarize` is given) only print changes. It used to work that way, but I changed it to put in `-` to indicate no subs and no length change and didn't update the checking as to whether to print something at all. You asked about that output last night and if it printed even if there were no changes.

terrycojones committed:
Moved some functionality from bin/annotate-genome.py into sars2seq/annotate.py so it can be imported and used from elsewhere.

LeonieMei commented on bin/annotate-genome.py:
```diff
@@ -0,0 +1,73 @@
+
+
+if __name__ == "__main__":
+    sys.exit(main(makeParser().parse_args()))
```
Is there a reason for not using an exit in `main`?

terrycojones replied:
No. This is just a pattern I see quite often.

LeonieMei commented on sars2seq/annotate.py:
```diff
@@ -0,0 +1,154 @@
+    @return: A C{dict} with C{str} feature names as keys. See the 'result'
+        variable below.
+    """
+    alignment = SARS2Alignment(genome, features, aligner=aligner)
```
I don't quite understand why the class is called SARS2Alignment? Or is this just a remnant?

terrycojones replied:
It's a remnant.

LeonieMei commented on sars2seq/features.py:
```diff
@@ -236,20 +252,95 @@ def _initializeFromGenBankRecord(self, record: SeqIO.SeqRecord) -> None:
+
+        for offset in range(len(self.reference)):
+            if offset in annotatedOffsets:
+                if start is not None:
```
I don't quite understand this, from a first glance it looks to me like you are adding an annotated region to a set of unannotated regions? But I guess this is more complicated and I just don't understand what is going on.

terrycojones replied:
If start is not None, it means we have (previously) seen the beginning of an unannotated region. So when we next see an offset that is in an annotated region, we now we have reached the end of an unannotated region.  Is that any clearer?

I need to break this function out and write tests for it - sorry!

LeonieMei replied:
Ok yes I think I understand it now, thanks!

LeonieMei reviewed (commented):
looks good!

terrycojones committed:
Black.
